File: aviva/aviva.go
/**

Get aviva balance

*/

package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/jessevdk/go-flags"
	"github.com/plord12/webscrapers/utils"

	"github.com/playwright-community/playwright-go"
)

type Options struct {
	Headless        bool   `short:"e" long:"headless" description:"Headless mode" env:"HEADLESS"`
	Username        string `short:"u" long:"username" description:"Aviva username" env:"AVIVA_USERNAME" required:"true"`
	Password        string `short:"p" long:"password" description:"Aviva password" env:"AVIVA_PASSWORD" required:"true"`
	Otppath         string `short:"o" long:"otppath" description:"Path to file containing one time password message" default:"otp/aviva" env:"OTP_PATH"`
	Otpcommand      string `short:"c" long:"otpcommand" description:"Command to get one time password" env:"OTP_COMMAND"`
	Otpcleancommand string `short:"l" long:"otpcleancommand" description:"Command to clean previous one time password" env:"OTP_CLEANCOMMAND"`
}

var options Options
var parser = flags.NewParser(&options, flags.Default)

func main() {

	// parse flags
	//
	_, err := parser.Parse()
	if err != nil {
		os.Exit(0)
	}

	// clean from any previous run
	//
	utils.CleanOTP(options.Otpcleancommand, options.Otpcleancommand)

	// setup
	//
	page := utils.StartCamoufox(options.Headless)
	defer utils.Finish(page)

	// Aviva can be really slow
	page.SetDefaultTimeout(60000.0)

	// main page & login
	//
	log.Printf("Starting login\n")
	_, err = page.Goto("https://www.direct.aviva.co.uk/MyAccount/login", playwright.PageGotoOptions{WaitUntil: playwright.WaitUntilStateDomcontentloaded})
	if err != nil {
		panic(fmt.Sprintf("could not goto url: %v", err))
	}

	// dismiss pop-up
	//
	page.GetByText("Essential cookies only", playwright.PageGetByTextOptions{Exact: playwright.Bool(true)}).Click(playwright.LocatorClickOptions{Timeout: playwright.Float(2000.0)})

	log.Printf("Logging in\n")
	// <input aria-required="True" autocomplete="off" class="a-textbox" data-qa-textbox="username" data-val="true" data-val-required="Please enter your username" id="username" maxlength="50" name="username" type="text" value="">
	err = page.Locator("#username").Fill(options.Username)
	if err != nil {
		panic(fmt.Sprintf("could not get username: %v", err))
	}
	// <input aria-required="True" autocomplete="off" class="a-textbox" data-qa-textbox="password" data-val="true" data-val-required="Please enter your password" id="password" maxlength="300" name="password" type="password">
	err = page.Locator("#password").Fill(options.Password)
	if err != nil {
		panic(fmt.Sprintf("could not get password: %v", err))
	}
	// <input id="loginButton" name="loginButton" class="a-button a-button--primary dd-data-link" data-dd-group="myAvivaLogin" data-dd-loc="login" data-dd-link="login" type="submit" value="Log in" data-qa-button="submitForm">
	err = page.Locator("#loginButton").Click()
	if err != nil {
		panic(fmt.Sprintf("could not click: %v", err))
	}

	// dismiss pop-up
	//
	// <button id="onetrust-accept-btn-handler">Accept all cookies</button>
	page.GetByText("Accept all cookies", playwright.PageGetByTextOptions{Exact: playwright.Bool(true)}).Click(playwright.LocatorClickOptions{Timeout: playwright.Float(2000.0)})

	// attempt to fetch one time password if needed
	//
	utils.FetchOTP(options.Otpcommand)

	// check/poll if otp/aviva exists ... could be via the above command or pushed here elsewhere
	//
	otp := utils.PollOTP(options.Otppath)

	if otp != "" {
		log.Println("otp=" + string(otp))

		err = page.Locator("#factor").Fill(otp)
		if err != nil {
			panic(fmt.Sprintf("could not set otp: %v", err))
		}

		err = page.Locator("#VerifyMFA").Click()
		if err != nil {
			panic(fmt.Sprintf("could not click otp: %v", err))
		}
	} else {
		panic("could not get one time password")
	}

	// get balance
	//

	balance, err := page.Locator("[data-qa-text=total-plan-value]").TextContent()
	if err != nil {
		panic(fmt.Sprintf("failed to get balance: %v", err))
	}
	log.Println("balance=" + balance)
	fmt.Println(strings.NewReplacer("£", "", ",", "").Replace(balance))

	bufio.NewWriter(os.Stdout).Flush()
}

File: aviva/aviva.py
#!/usr/bin/env python3
#
# selenium script to get aviva balance
#
# AVIVA_USER=user AVIVA_PWD=password ./aviva.py
#

import time
import re 
import os 
import os.path
import sys
import platform
import undetected_chromedriver as uc
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait

# clean up
#
if 'OTP_CLEANCOMMAND' in os.environ:
    print("Running "+os.environ['OTP_CLEANCOMMAND'], file=sys.stderr)
    os.system(os.environ['OTP_CLEANCOMMAND'])
if os.path.exists('otp/aviva'):
  os.remove('otp/aviva')

# Note: use_subprocess=False fails on mac, should be set true
# Note: driver is assumed to be x86-64 on linux, so we copy locally installed chromedriver
#
if platform.system() == "Linux":
    driver = uc.Chrome(headless=False,use_subprocess=False,driver_executable_path=os.environ['HOME']+"/.local/share/undetected_chromedriver/chromedriver_copy")
else:
    driver = uc.Chrome(headless=False,use_subprocess=True)

try:
    driver.get('https://www.direct.aviva.co.uk/MyAccount/login')

    WebDriverWait(driver, 10).until(EC.element_to_be_clickable((By.CSS_SELECTOR,'#onetrust-accept-btn-handler'))).click()

    WebDriverWait(driver, 10).until(EC.element_to_be_clickable((By.CSS_SELECTOR,'#loginButton')))
    driver.find_element(By.CSS_SELECTOR,'#username').send_keys(os.environ['AVIVA_USERNAME'])
    driver.find_element(By.CSS_SELECTOR,'#password').send_keys(os.environ['AVIVA_PASSWORD'])
    driver.find_element(By.CSS_SELECTOR,'#loginButton').click()

    time.sleep(5)

    # otp
    count = 100
    while not os.path.isfile('otp/aviva') and count > 0:
        if 'OTP_COMMAND' in os.environ:
            print("Running "+os.environ['OTP_COMMAND'], file=sys.stderr)
            os.system(os.environ['OTP_COMMAND'])
        time.sleep(1)
        count = count-1
    f = open('otp/aviva', 'r')
    otp = re.findall("code is: ([0-9]*).", f.read())[0]
    f.close()
    print("otp="+otp, file=sys.stderr)
    os.remove('otp/aviva')

    driver.find_element(By.CSS_SELECTOR,'#factor').send_keys(otp)
    driver.find_element(By.CSS_SELECTOR,'#VerifyMFA').click()
    time.sleep(5)

    driver.find_element(By.PARTIAL_LINK_TEXT, 'Details').click()
    time.sleep(20)

    x = re.search('\"yourPensionValue\"[^0-9]*([0-9,.]*)[^0-9]*', driver.page_source) 

    print(x[1].replace(',', '')) 

except:
    driver.save_screenshot('debug-aviva-'+os.environ['AVIVA_USERNAME']+'.png')
